app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,160 @@
+        val removed = keys.filter {
+            val pos = byKey.remove(it) ?: return@filter false
+            byTile[getTilePosFor(pos)]?.remove(it)
+            true
+        }
+        onKeysRemoved(removed)
+    }
+
+    // returns whether the key was put in cache
+    fun putIfTileExists(key: T, position: LatLon): Boolean {
+        val previousPosition = byKey[key]
+        // in cache already
+        if (previousPosition != null) {
+            // but moved -> remove
+            if (previousPosition != position) remove(key)
+            else return true
+        }
+        val tileInCache = byTile[getTilePosFor(position)]
+        if (tileInCache  != null) {
+            tileInCache.add(key)
+            byKey[key] = position
+            return true
+        }
+        return false
+    }
+
+    // returns keys that were not put to cache because they are not in fully contained tiles
+    // contrary to get(bbox), here the bbox may be larger than what can be put in cache
+    fun replaceAllInBBox(entries: Collection<Pair<T, LatLon>>, bbox: BoundingBox): List<T> {
+        val tiles = bbox.asListOfEnclosingTilePos()
+        val completelyContainedTiles = tiles.filter { it.asBoundingBox(tileZoom).isCompletelyInside(bbox) }
+        if (tiles != completelyContainedTiles)
+            Log.i(TAG, "replacing tiles not completely contained in BBox")
+        val tilesToFetch = if (completelyContainedTiles.size > maxTiles) {
+            Log.i(TAG, "trying to replacing more tiles than fit in cache, ignoring some tiles")
+            completelyContainedTiles.subList(0,maxTiles-1)
+        } else
+            completelyContainedTiles
+        val ignoredKeys = replaceAllInTiles(entries, tilesToFetch)
+
+        // trim only when bbox!
+        trim()
+        return ignoredKeys
+    }
+
+    // may add tiles, but does not call trim()
+    private fun replaceAllInTiles(entries: Collection<Pair<T, LatLon>>, tiles: Collection<TilePos>): List<T> {
+        // create / replace tiles
+        tiles.forEach { tile ->
+            byTile.remove(tile)?.let { byKey.keys.removeAll(it) }
+            byTile[tile] = hashSetOf()
+        }
+        if (tiles.size == 1) {
+            // shortcut: replace old tile
+            byTile[tiles.single()] = entries.unzip().first.toHashSet()
+            byKey.putAll(entries)
+            return emptyList()
+        } else {
+            // put only what is inside tiles
+            // and return what wasn't put
+            val ignoredKeys = mutableListOf<T>() // could also use sth like filter or mapNotNull instead of mutable list
+            val bboxByTile = tiles.associateWith { it.asBoundingBox(tileZoom) }
+            entries.forEach { (key, pos) ->
+                for (tile in tiles) {
+                    if (bboxByTile[tile]!!.contains(pos)) {
+                        byTile[tile]!!.add(key)
+                        byKey[key] = pos
+                        return@forEach
```

Comment:
I think you think the `return@forEach` does something else than it actually does. (Because of these kind of things, I usually prefer `for (x in y)`).

See https://pl.kotl.in/FA0r8L_PL

Reply:
What I want to achieve here is that if the position is contained in a tile, we break the `for (tile in tiles)` loop without adding the `key` to `ignoredKeys`, i.e. it's used as `continue` to go to the next entry in the `forEach` loop.

Reply:
Ah okay, then this looks like it works as intended

Reply:
anyway, changed to for loop now to be clear